### PR TITLE
Maintaining version in StoreFindToken during ser and deser

### PR DESCRIPTION
**Pull request by nsivabalan**
Currently, StoreFindToken doesn't have a version strictly associated with it and hence there are chances that a token that was serialized in V1, while reading it back will take up version 2 implicitly.

This patch adds version field to StoreFindtoken so that serialization is based on the version of the token and de-serialization maintains the same version. Any new StoreFindToken created (using public constructors) will take up the latest version(i.e. Version 2). This will ensure that any token read in version X will be written in a version <=X and never in a version > X.
Added unit tests to verify that version is maintained during serialization and deserialization as expected. 

More details: https://github.com/linkedin/ambry/issues/551

SLA: 10 mins
Reviewers: @pnarayanan @vgkholla 
Coverage: (running StoreFindTokenTest)
StoreFindToken	100% (3/ 3)	90.9% (20/ 22)	95.5% (169/ 177)
Verified that all new lines are covered except for few error cases (unknown store find token version and type) 

Built and coding style applied

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 1]
```diff
@@ -93,11 +95,60 @@
...
+    this(Type.JournalBased, offset, null, sessionId, incarnationId, inclusive, VERSION_2);
   }
 
-  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId,
-      boolean inclusive) {
+  /**
+   * Uninitialized token with the given version. Refers to the starting of the log.
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(short version) {
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: 7 constructors is really too many. Could we clean this up?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 2]
```diff
@@ -245,52 +303,102 @@ public long getBytesRead() {
...
+        bufWrap.putShort((short) type.ordinal());
+        // add offset
+        bufWrap.put(offsetBytes);
+        // add StoreKey
+        bufWrap.put(storeKeyBytes);
+        break;
+      case VERSION_2:
+        offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        byte[] incarnationIdBytes = incarnationId != null ? incarnationId.toString().getBytes() : ZERO_LENGTH_ARRAY;
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: If this is null, fail to serialize and throw an error. This should never be null in fact, and that should be ensured at the time of construction.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 1 above)
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: yes. 3 are public (Unitialized, journal and index). 3 are private ones so that deserialization takes in a version apart from usual args in public constructors. And one private constructor which is called from every other constructor. We can make the deserialization calls to explicitly call the last private with default values, but thought this would mimic how older version tokens are created

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 2 above)
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: added

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 1 above)
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: +1

7 is too much. Three public and one private (like before) is preferred.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 3]
```diff
@@ -65,12 +65,14 @@
   private final UUID incarnationId;
   // refers to the bytes read so far (from the beginning of the log)
   private long bytesRead;
+  // refers to the version of the StoreFindToken
+  private short version;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: `final`?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 4]
```diff
@@ -93,11 +95,60 @@
...
+   * @param key The {@link StoreKey} which the token refers to. Index segments are keyed on store keys and hence
+   * @param sessionId the sessionId of the store that this token referes to
+   * @param incarnationId the incarnationId of the store that this token refers to
+   * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
+   *                  {@code false} otherwise
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId, boolean inclusive,
+      short version) {
+    if (version != VERSION_0 && version != VERSION_1 && version != VERSION_2) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: this is a private constructor - when can this situation happen? When constructing from constructors, the version is the current version by default. When deserializing, a wrong version would have failed deserialization.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 5]
```diff
@@ -235,6 +289,10 @@ boolean getInclusive() {
     return inclusive == (byte) 1;
   }
 
+  short getVersion() {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: doc

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 6]
```diff
@@ -245,52 +303,105 @@ public long getBytesRead() {
...
-      bufWrap.put(sessionIdBytes);
-      // add offset
-      bufWrap.put(offsetBytes);
-      if (type == Type.JournalBased) {
-        bufWrap.put(getInclusive() ? (byte) 1 : (byte) 0);
-      } else if (type == Type.IndexBased) {
+    byte[] buf = null;
+    switch (version) {
+      case VERSION_0:
+        int OFFSET_SIZE = 8;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: all caps are not appropriate here.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 7]
```diff
@@ -245,52 +303,105 @@ public long getBytesRead() {
...
+        // add offset
+        bufWrap.put(offsetBytes);
+        // add StoreKey
+        bufWrap.put(storeKeyBytes);
+        break;
+      case VERSION_2:
+        offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        if (type != Type.Uninitialized && incarnationId == null) {
+          throw new IllegalStateException("IncarnationId cannot be null for Journal or Index based token ");
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: this can verified much before getting here? i.e at construction time.

Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: I did give it a thought. Before this patch, any token (of older verions) on deserialization will take up version 2 and hence can't have this check. Hence, we don't explicitly check for null incarnationId during construction. With this patch, that is not valid anymore. But in our tests(serDeTest), we create older type versions using normal constructor and we use private methods to serialize in particular version. If we have the null check, then might have to come up with a private method which takes in arguments similar to StoreFindToken constructors and go about it. Want to keep the changes minimal as this is blocking our builds.

Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: I'm not sure I follow completely but is there any problem in doing 
```
if(version == VERSION_2 && type != UNINITIALIZED && incarnationId == null) {
  // throw
}
```
at construction time?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 8]
```diff
@@ -93,11 +95,60 @@
...
+   *                  {@code false} otherwise
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId, boolean inclusive,
+      short version) {
+    if (version != VERSION_0 && version != VERSION_1 && version != VERSION_2) {
+      throw new IllegalArgumentException("Unsupported version " + version + " for StoreFindToken ");
+    }
     if (!type.equals(Type.Uninitialized)) {
       if (offset == null || sessionId == null) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: can whatever I recommended [here](https://github.com/linkedin/ambry/pull/553/files/2b45a8b2d2f7a9b18ccb040135fc1cbd0ab4b5ab#r102640486) fit as another `else if` here?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 7 above)
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: yes, we can't do it easily. Check for serDeTest() in StoreFindTokenTest(). We do create tokens in old format and we have a private method to return the serialized form for the given version. With the extra chuck for null incarnationId, we can do that. Have to introduce another method which takes in all arguments and the required version to return the serialized form of it.

**Comment by nsivabalan** [comment]
@vgkholla @pnarayanan : done addressing comments

**Comment by nsivabalan** [comment]
@pnarayanan @vgkholla let me know if you have any more comments

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 9]
```diff
@@ -93,17 +95,31 @@
...
+    this(Type.JournalBased, offset, null, sessionId, incarnationId, inclusive, VERSION_2);
   }
 
-  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId,
-      boolean inclusive) {
+  /**
+   * Instantiating {@link StoreFindToken}
+   * @param type the {@link Type} of the token
+   * @param offset the offset that this token refers to
+   * @param key The {@link StoreKey} which the token refers to. Index segments are keyed on store keys and hence
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: minor: "and hence"? Sentence looks incomplete.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 10]
```diff
@@ -93,17 +95,31 @@
...
+   */
+  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId, boolean inclusive,
+      short version) {
     if (!type.equals(Type.Uninitialized)) {
       if (offset == null || sessionId == null) {
         throw new IllegalArgumentException("Offset [" + offset + "] or SessionId [" + sessionId + "] cannot be null");
       } else if (type.equals(Type.IndexBased) && key == null) {
         throw new IllegalArgumentException("StoreKey cannot be null for an index based token");
       }
+      if(version == StoreFindToken.VERSION_2 && incarnationId == null){
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: minor: I'd say `if version >= VERSION_2` might be a better check

Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: Don't need the `StoreFindToken.` qualifier

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 11]
```diff
@@ -245,52 +274,102 @@ public long getBytesRead() {
...
+            offsetSize + startOffsetSize + storeKeyBytes.length;
+        buf = new byte[size];
+        ByteBuffer bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(version);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add offset for journal based token
+        bufWrap.putLong((type == Type.JournalBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: can offset be null if type is journal based?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 12]
```diff
@@ -245,52 +274,102 @@ public long getBytesRead() {
...
+            : UNINITIALIZED_OFFSET);
+        // add index start offset for Index based token
+        bufWrap.putLong((type == Type.IndexBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: same question as above

File: ambry-store/src/test/java/com.github.ambry.store/IndexTest.java [hunk 13]
```diff
@@ -52,6 +52,15 @@
   private final boolean isLogSegmented;
   private final File tempDir;
   private final CuratedLogIndexState state;
+  private static final StoreKeyFactory STORE_KEY_FACTORY;
+
+  static {
+    try {
+      STORE_KEY_FACTORY = Utils.getObj("com.github.ambry.store.MockIdFactory");
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: you can use the `STORE_KEY_FACTORY` in `CuratedLogIndexState` (make it package private if required)

File: ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java [hunk 14]
```diff
@@ -140,34 +140,25 @@ public void serDeTest() throws IOException {
       doSerDeTest(new StoreFindToken(), StoreFindToken.VERSION_0, StoreFindToken.VERSION_1, StoreFindToken.VERSION_2);
 
       // Journal based token
-      // incarnationId cannot be null for VERSION_2
-      doSerDeTest(new StoreFindToken(offset, sessionId, null, false), StoreFindToken.VERSION_0,
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java]: why are these tests no longer required?

File: ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java [hunk 15]
```diff
@@ -226,9 +217,22 @@ private void doSerDeTest(StoreFindToken token, Short... versions) throws IOExcep
...
       DataInputStream stream = getSerializedStream(token, version);
       StoreFindToken deSerToken = StoreFindToken.fromBytes(stream, STORE_KEY_FACTORY);
+      assertEquals("Stream should have ended ", 0, stream.available());
+      assertEquals("Version mismatch for token ", version.shortValue(), deSerToken.getVersion());
       compareTokens(token, deSerToken);
       assertEquals("SessionId does not match", token.getSessionId(), deSerToken.getSessionId());
+      if (version == StoreFindToken.VERSION_2) {
+        assertEquals("IncarnationId mismatch ", token.getIncarnationId(), deSerToken.getIncarnationId());
+      }
+      stream = new DataInputStream(new ByteBufferInputStream(ByteBuffer.wrap(deSerToken.toBytes())));
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java]: what is the intent behind doing this a second time? i.e. what is the failure scenario when the preceding part does not fail, but this does. During serialization?

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 11 above)
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: offset cannot be null in both cases. We throw IllegalArgumentException in the constructor. Have removed the null check here.

File: ambry-store/src/test/java/com.github.ambry.store/IndexTest.java (hunk 13 above)
Comment (nsivabalan) [inline ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: sure

File: ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java (hunk 14 above)
Comment (nsivabalan) [inline ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java]: because, the previous tests inherently tested this path. In other words, even though we create StoreFindToken with incarnationId, during serialization in version 0 and version 1 incarnation Id is not accounted for. Also, have added tests in constructionErrorCasesTest for null incarnationId.

File: ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java (hunk 15 above)
Comment (nsivabalan) [inline ambry-store/src/test/java/com.github.ambry.store/StoreFindTokenTest.java]: Sorry. I tried to think through why I added it, but couldn't find a reason. Will remove it for now. If you can think of a reason why it could fail, do let me know.

**Comment by vgkholla** [comment]
LGTM. Please verify formatting and build on OSx and Linux. Also add unit testing details to the PR description.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: could you drop this `incarnationId != null` check? It has the unintentional connotation of null incarnation id being okay for version 2.

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java [hunk 16]
```diff
@@ -93,17 +96,31 @@
...
+   */
+  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId, boolean inclusive,
+      short version) {
     if (!type.equals(Type.Uninitialized)) {
       if (offset == null || sessionId == null) {
         throw new IllegalArgumentException("Offset [" + offset + "] or SessionId [" + sessionId + "] cannot be null");
       } else if (type.equals(Type.IndexBased) && key == null) {
         throw new IllegalArgumentException("StoreKey cannot be null for an index based token");
       }
+      if (version >= VERSION_2 && incarnationId == null) {
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: Let us make this check `==`, (> VERSION_2 by itself is an illegal argument really).

Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: done

**Comment by nsivabalan** [comment]
@vgkholla @pnarayanan : have addressed all your comments

File: ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java (hunk 2 above)
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java]: UnInitialized token can have null incarnationId and hence we can't drop this. We have two options here. Either leave it as is. Or ensure that even Uninitialized token has incarnationId set.

**Comment by vgkholla** [comment]
@nsivabalan unit test coverage?

**Comment by pnarayanan** [comment]
@nsivabalan any reason why this branch is called "unBlockBuildFailure"?

**Comment by nsivabalan** [comment]
@pnarayanan : Sorry that was by mistake. I should have named it as StoreFindTokenV2Fix.

**Comment by pnarayanan** [comment]
@nsivabalan has tested the change, merging.
